Add open.py

diveshuttam opened the pull request:
closes #31

diveshuttam committed:
Add open.py

dashohoxha commented:
Why do you name the file `eopen.py`?
You can name the function `open_file()` in case that it conflicts with some other function, it doesn't have to be the same with the name of the file.

dashohoxha commented on src/gpg/eopen.py:
```diff
@@ -0,0 +1,72 @@
+import time
+import textwrap
+
+DEBUG= True if (os.environ['DEBUG']=='yes') else False
```
Maybe it can be simplified like this:
`DEBUG = (os.environ['DEBUG'] == 'yes')`
Or maybe you don't need an extra variable since you can use the test condition above.

dashohoxha commented on src/gpg/eopen.py:
```diff
@@ -0,0 +1,72 @@
+                exit(3)
+                
+        while(os.path.exists(output_file_path)):
+            print("File {output} already exists".format(output=output_file_path))
```
Making this check in the bash script is simpler. Just fail with a proper message if output file already exists.

diveshuttam replied:
The gpg command that I replaced had this behavior, it asked to override or enter a new file name if the file already exists.
If I understood correctly you are suggesting to move this checking functionality to bash script before calling this python script??

dashohoxha replied:
> If I understood correctly you are suggesting to move this checking functionality to bash script before calling this python script??

This is right.

> The gpg command that I replaced had this behavior, it asked to override or enter a new file name if the file already exists.

We don't have to duplicate the behavior of `gpg`.
In this case I think that it is more reasonable that we just inform the user that a file of that name already exists, and stop. The user can rename or delete the existing file and try the command again.

dashohoxha replied:
Also don't forget to apply good pythonic style to this script too.

diveshuttam replied:
Yup I'll add it just in a while.

dashohoxha commented on src/gpg/eopen.py:
```diff
@@ -0,0 +1,72 @@
+
+    except BaseException:
+        if(DEBUG):
+            raise
```
Try to finish this task as soon as possible: https://github.com/EasyGnuPG/pgpg/issues/54
because it may also help you with development.

diveshuttam replied:
opened #62 for the same.

dashohoxha commented:
Are you checking that the automated test scripts still work with the changes that you are making?

diveshuttam commented:
I did try this today for the seal command but couldn't find the cause for failing of those tests, they were failing because no keys were there in the trash directory debugged it for a while but couldn't figure out the exact reason. 
Though I know that I have added some prompts so changes are needed but they are not the cause as of now. Am working on this on priority will add them in just a while.By the time you can suggest changes on other aspects.

diveshuttam commented:
>Why do you name the file eopen.py?
You can name the function open_file() in case that it conflicts with some other function, it doesn't have to be the same with the name of the file.

I mistakenly had the ruins of the pure python version in mind somewhere while implementing this. In that case would have imported this file as a module and might have conflicted. Not required now, have updated it.

dashohoxha commented:
> I did try this today for the seal command but couldn't find the cause for failing of those tests, they were failing because no keys were there in the trash directory debugged it for a while but couldn't figure out the exact reason.

Do the tests work on `egpg`? If yes then most probably they are broken because of your changes.
Try the `-d` option with test scripts, in order to get more detailed output, which may help you to debug.

diveshuttam commented:
>Do the tests work on egpg? 

Yes most test work on the gnupg-2.2 branch of egpg leaving a few which require havaged

>If yes then most probably they are broken because of your changes.

Yup, I also think it is and am working on it.

>Try the -d option with test scripts, in order to get more detailed output, which may help you to debug.

I had tried it already but the -d only showed that I didn't have valid keys.
UPDATE:
BTW as you suggested this again, I compared the output of both (test cases of pgpg and egpg2.2). I copied the test cases of EasyGnuPG/egpg to my local pgpg repo and they are also working on pgpg now. I think either 1) we have old test cases in this repo.  2) I might have messed something up with my local version.

I'll continue this in the morning.

diveshuttam committed:
Merge branch 'gnupg-2.2' into issue31

update local branch

diveshuttam committed:
update for pep8

diveshuttam committed:
Move file checking functionality to bash file.

Not checking in open.py now instead in open.sh
The code in the bash script is similar to seal.sh

diveshuttam committed:
Make test for open pass

diveshuttam commented:
Hi I have updated these according to PEP8. Tests are passing as well as I have moved the checking functionality to open.sh file. Please review.
Also, The code I have added for checking if output file exists in open.sh is similar to the one in seal.sh for consistency.

dashohoxha commented on src/gpg/open.py:
```diff
@@ -0,0 +1,60 @@
+                exit(3)
+
+        with open(output_file_path, "wb") as nfile:
+            nfile.write(plaintext)
```
Can these 2 lines be written like this?
```
nfile = open(output_file_path, "wb")
nfile.write(plaintext)
```

If yes, then the `with` construct seems a bit redundant in this case.

diveshuttam replied:
Similar but not exact they just differ by a close statement. `with` constructs guarantees that files are closed after the with gets over. It's just a nice way to depict that a particular resource will be closed after this and will only be used in `with`'s boundaries. 
Yes for small no. of operations the one you suggested (and adding a close at the end) is also fine but this helps in larger ones. And I prefer it as it is one line less :-) .

dashohoxha reviewed (approved):
It seems Ok to me.

sourabhtk37 commented on src/gpg/open.py:
```diff
@@ -0,0 +1,60 @@
+    for signature in verify_result.signatures:
+        user = c.get_key(signature.fpr).uids[0].uid
+        fpr = signature.fpr
+        signed_time = time.ctime(signature.timestamp)
```
Additional variables(fpr, signed_time) seems unnecessary to me only for the purpose of formatting, don't you think?
In GPGME docs, we have: 
```
format(c.get_key(sign.fpr).uids[0].uid,
       sign.fpr, time.ctime(sign.timestamp)))

diveshuttam replied:
Ok will update in another commit. I took this from verify.py. 
In GPGME docs, these were being used with {0}, {1}, {2} so they were not looking fine to me as they didn't signify their intent. So, In verify.py I first declared these with proper names and used these with {0}, {1}, {2} which seems fine to me. Later I updated formatting with named variables in so this is a remains from there.  Will remove these and use something like
```
.
.
.
               '''.format(user=c.get_key(signature.fpr).uids[0].uid,
                             fingerprint=signature.fpr,
                             time=time.ctime(signature.timestamp))
```
This will signify the intent and be concise as well.
